### ISSUE TITLE
ASP.NET Core 2.0 Web API Middleware - HttpContext Response OnStarting is not called

### ISSUE THREAD

**Issue by frankyvij** [aws/aws-sdk-net#857]
I have created a middleware to parse the response body and change the status code based on response. I have tested on my local, and it works fine. However, when I deployed the solution to AWS Lambda, it did not work. I have put some logs to see where does the execution goes; but have been unsuccessful.

```
public class ResponseFailureMiddleware
    {
        private readonly RequestDelegate _next;

        public ResponseFailureMiddleware(RequestDelegate next)
        {
            _next = next;
        }

        public async Task Invoke(HttpContext context)
        {
            LambdaLogger.Log("ResponseFailureMiddleware Invoked");

            context.Response.OnStarting(state => {
                LambdaLogger.Log("ResponseFailureMiddleware OnStarting");
                if (context.Response != null && context.Response.ContentType != null && context.Response.ContentType.Contains("application/json"))
                {
                    try
                    {
                        LambdaLogger.Log("Reading response body");
                        using (Stream stream = context.Response.Body)
                        {
                            stream.Seek(0, SeekOrigin.Begin);
                            var body = JsonConvert.DeserializeObject<dynamic>(new StreamReader(stream).ReadToEnd());

                            LambdaLogger.Log($"Read response body. API returned success: {body.success}");
                            // If success is false, then change the response status code to 400
                            if (body.success == false)
                            {
                                var httpContext = (HttpContext)state;
                                httpContext.Response.StatusCode = 400;
                                LambdaLogger.Log($"Setting response status code to: {400}");
                            }
                        }
                    }
                    catch (Exception)
                    {
                        Console.WriteLine("Unable to parse the response. Skipping middleware logic.");
                    }
                }
                return Task.FromResult(0);
            }, context);

            await _next(context);
        }
```

Could it be that this event is not suppose to be called on AWS Lambda  @environment? If so, then is there any other event that I can capture in my code to transform the response? Thanks.

**Comment by frankyvij** [comment]
As an alternative to above, I am using the code below to transform the response:

     ```
   public async Task Invoke(HttpContext context)
        {
            LambdaLogger.Log("ResponseFailureMiddleware Invoked");

            var originBody = context.Response.Body;

            var newBody = new MemoryStream();

            context.Response.Body = newBody;

            await _next(context);

            if (context.Response != null && context.Response.ContentType != null && context.Response.ContentType.Contains("application/json"))
            {
                try
                {
                    LambdaLogger.Log("Reading response body");
                    newBody.Seek(0, SeekOrigin.Begin);
                    string json = new StreamReader(newBody).ReadToEnd();
                    var body = JsonConvert.DeserializeObject<dynamic>(json);

                    LambdaLogger.Log($"Read response body. API returned success: {body.success}");
                    // If success is false, then change the response status code to 400
                    if (body.success == false)
                    {
                        context.Response.StatusCode = 400;
                        LambdaLogger.Log($"Setting response status code to: {400}");
                    }

                    context.Response.Body = originBody;

                    await context.Response.WriteAsync(json);
                }
                catch (Exception)
                {
                    Console.WriteLine("Unable to parse the response. Skipping middleware logic.");
                }
            }
        }
```

**Comment by sstevenkang** [comment]
What did you mean by "it did not work"?  Could you elaborate on what's logged in cloudwatch or what error response you get from Lambda?

**Comment by frankyvij** [comment]
Oh, I think I did not put it correctly. What I mean is that OnStarting delegate of HttpContext.Response is not invoked when API is hosted in Lambda. However, the same event is fired when I test on my local. There is no error caused by this; it just that it simply does not invoke the delegate just before the headers are sent to the client.

Also, the only log I see in CloudWatch around this is (since OnStarting is never invoked):
**ResponseFailureMiddleware Invoked**

I will do a sample application and upload it for your test on your local system and lambda.

**Comment by frankyvij** [comment]
Attached is the sample application which demonstrates the issue I mentioned above.

- Download and run the solution on your local
- Send a GET request (in my case I did: /api/values/5) through any REST client and notice that OnStarting delegate gets invoked in the middleware. (also note that there is a log which should appear in CloudWatch when the same thing is executed in Lambda)
- Publish the solution in AWS
- Send the same GET on the new URL
- Notice that in CloudWatch there is no log from OnStarting delegate

Here's the CloudWatch log from my test:
```
[Warning] Microsoft.AspNetCore.DataProtection.Repositories.EphemeralXmlRepository: Using an in-memory repository. Keys will not be persisted to storage. 
[Warning] Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager: Neither user profile nor HKLM registry available. Using an ephemeral key repository. Protected data will be unavailable when application exits. 
[Information] Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager: Creating key {48100f92-957f-4ca2-b124-a7aee046ad3a} with creation date 2018-02-06 01:37:47Z, activation date 2018-02-06 01:37:47Z, and expiration date 2018-05-07 01:37:47Z. 
[Warning] Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager: No XML encryptor configured. Key {48100f92-957f-4ca2-b124-a7aee046ad3a} may be persisted to storage in unencrypted form. 
START RequestId: 5543201e-0ade-11e8-a90e-65f9554675e9 Version: $LATEST
Incoming GET requests to /api/values/5
ASP.NET Core Request PathBase: /Prod, Path: /api/values/5
[Information] Microsoft.AspNetCore.Hosting.Internal.WebHost: Request starting GET https://xi1kn1cz91.execute-api.ap-southeast-1.amazonaws.com/Prod/api/values/5 
Middleware Invoked
[Information] Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker: Executing action method ResponseMiddlewareTest.Controllers.ValuesController.Get (ResponseMiddlewareTest) with arguments (5) - ModelState is Valid 
[Information] Microsoft.AspNetCore.Mvc.Internal.ObjectResultExecutor: Executing ObjectResult, writing value Microsoft.AspNetCore.Mvc.ControllerContext. 
[Information] Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker: Executed action ResponseMiddlewareTest.Controllers.ValuesController.Get (ResponseMiddlewareTest) in 198.7709ms 
[Information] Microsoft.AspNetCore.Hosting.Internal.WebHost: Request finished in 962.0843ms 0 text/plain; charset=utf-8 
Response Base 64 Encoded: False
END RequestId: 5543201e-0ade-11e8-a90e-65f9554675e9
REPORT RequestId: 5543201e-0ade-11e8-a90e-65f9554675e9	Duration: 1652.77 ms	Billed Duration: 1700 ms Memory Size: 256 MB	Max Memory Used: 58 MB	
```

[ResponseMiddlewareTest.zip](https://github.com/aws/aws-sdk-net/files/1697410/ResponseMiddlewareTest.zip)

**Comment by normj** [comment]
I see we need to implement the OnStarting method in our IHttpResponseFeature. Right now it is just a stub. I'll work on getting those properly implemented. 

https://github.com/aws/aws-lambda-dotnet/blob/6bb119c3fec64d9afd8580b7184ed92615ece28f/Libraries/src/Amazon.Lambda.AspNetCoreServer/Internal/InvokeFeatures.cs#L141

**Comment by normj** [comment]
I released version 2.0.1 of `Amazon.Lambda.AspNetCoreServer` that now implements the OnStarting and OnComplete callbacks. Keep in mind for the Lambda context OnStarting and OnComplete are called when the response to API Gateway is being created. Not when the results are actually sent to the client.
